app/parser/metadata_parser.py

```diff
@@ -23,67 +25,100 @@ def id_generator():
 }
 
 
 class MetadataParser(object):
 
     @staticmethod
-    def is_valid(token):
-        for key, constant in metadata_constants.items():
-            if constant.mandatory and key not in token:
-                return False, key
-        return True, ""
-
-    @staticmethod
-    def build_metadata(token):
+    def parse(metadata_to_check):
+        parsed = {}
         try:
-            metadata = {}
-            # loop around all the constants and add them as attributes of the metadata object
-            for attr_name, constant in metadata_constants.items():
-                logger.debug("MetadataParser adding attr %s", attr_name)
-                if attr_name in token:
-                    value = token[attr_name]
-                    attr_value = constant.parser(value)
+            for key, field in metadata_fields.items():
+                logger.debug("MetadataParser adding attr %s", key)
+                if key in metadata_to_check:
+                    value = metadata_to_check[key]
+                    attr_value = field.parse(value)
                     logger.debug("with value %s", attr_value)
-                elif constant.mandatory:
-                    logger.warning("Missing constant value for %s", constant.claim_id)
-                    raise ValueError("Missing constant value %s".format(constant.claim_id))
+                elif field.mandatory:
+                    logger.warning("Missing field value for %s", key)
+                    raise ValueError("Missing field value %s".format(key))
                 else:
-                    if constant.generator:
-                        logger.debug("Generating value for %s", attr_name)
-                        attr_value = constant.generator()
-                    else:
-                        logger.debug("No value provide for %s but this is not mandatory, setting to None", attr_name)
-                        attr_value = None
-                metadata[attr_name] = attr_value
-            return metadata
+                    logger.debug("Generating value for %s", key)
+                    attr_value = field.generate()
+
+                parsed[key] = attr_value
         except (RuntimeError, ValueError, TypeError) as e:
             logger.error("Unable to parse Metadata")
             logger.exception(e)
             raise InvalidTokenException("Incorrect data in token")
+
+        return parsed
+
+    @staticmethod
+    def is_valid(token):
+        for key, field in metadata_fields.items():
+            if field.mandatory and key not in token:
+                return False, key
+        return True, ""
+
+
+def serialise_metadata(metadata):
+    # Strip all 'none' values
+    metadata = {k: v for k, v in metadata.items() if v is not None}
+    serialised_metadata = metadata
+    date_fields = [(k, metadata[k]) for k, v in metadata_fields.items() if v.parser == iso_8601_data_parser and k in metadata]
```

Comment:
Not with the current implementation - the metadata parse method returns a dict, thus what is finally passed to the final encode_metadata is a datetime object or string for each key, rather than a metadatafield which could be serialised in the way you describe. 

The flat dict is more desirable than the previous object and I think the metadatafield is really trying to do multiple jobs of validation/generation where we should probably be using a python module for the validation in this and other areas of the submitter. 

The commit above was really an interim step in making this class a collection of helper methods.